src/main/java/software/amazon/smithy/lsp/SmithyTextDocumentService.java

```diff
@@ -197,6 +225,97 @@ private String findToken(String path, Position p) throws IOException {
+                return new TextEdit(new Range(namespaceEnd, namespaceEnd), "\n" + line + trailingNewLine);
+            }
+        } else {
+            Position useBlockEnd = preamble.getUseBlockRange().getEnd();
+            return new TextEdit(new Range(useBlockEnd, useBlockEnd), "\n" + line + trailingNewLine);
+        }
+    }
+
+    private DocumentPreamble documentPreamble(List<String> lines) {
+
+        Range namespace = new Range(blankPosition, blankPosition);
+        Range useBlock = new Range(blankPosition, blankPosition);
+        Set<String> imports = new HashSet();
+
+        // First, we detect the namespace and use block in a very ugly way
+        boolean collectUseBlock = true;
+        int lineNumber = 0;
+        for (String line : lines) {
+            if (line.trim().startsWith("namespace ")) {
+                if (namespace.getStart() == blankPosition) {
+                    namespace.setStart(new Position(lineNumber, 0));
+                    namespace.setEnd(new Position(lineNumber, line.length() - 1));
+                }
+            } else if (line.trim().startsWith("use ")) {
+                String i = line.trim().split("use ", 2)[1].trim();
+                if (useBlock.getStart() == blankPosition) {
+                    imports.add(i);
+                    useBlock.setStart(new Position(lineNumber, 0));
+                    useBlock.setEnd(new Position(lineNumber, line.length()));
+                } else if (collectUseBlock) {
+                    imports.add(i);
+                    useBlock.setEnd(new Position(lineNumber, line.length()));
+                }
+            } else if (line.isBlank()) {
+                if (collectUseBlock)
+                    useBlock.setEnd(new Position(lineNumber, line.length()));
+            } else if (line.trim().startsWith("//")) {
+
+            } else {
+                collectUseBlock = false;
+            }
+
+            lineNumber++;
+        }
+
+        boolean blankSeparated = false;
+        // Next, we reduce the use block to the last use statement, ignoring newlines
+        // It's important to do so to make sure we don't multiply newlines
+        // unnecessarily.
+        if (useBlock.getStart() != blankPosition) {
+            while (lines.get(useBlock.getEnd().getLine()).isBlank()) {
+                blankSeparated = true;
+                int curLine = useBlock.getEnd().getLine();
+                useBlock.getEnd().setLine(curLine - 1);
+                useBlock.getEnd().setCharacter(lines.get(curLine - 1).length());
+            }
+        } else if (namespace.getStart() != blankPosition) {
+            int namespaceLine = namespace.getStart().getLine();
+            if (namespaceLine < lines.size() - 1) {
+                blankSeparated = lines.get(namespaceLine + 1).isBlank();
+            }
+        }
+
+        return new DocumentPreamble(namespace, useBlock, imports, blankSeparated);
+    }
```

Comment:
Given that we collect newlines (without knowing when to stop), we shrink back the use block by trimming last newlines.

It's also important to capture whether we trimmed _any_ newlines, because this will factor in the decision to insert (or not) a newline during auto-import.